### PR TITLE
ebos: implement partial support for the TUNING keyword

### PR DESCRIPTION
**Pull request by andlaus**
this only implements the parameters that are currently supported by `flow`.

large parts of this patch are due to a collaboration with @totto82. Thanks!

### CODE REVIEW

**Comment by andlaus** [comment]
jenkins build this please

**Comment by totto82** [comment]
The test failure is due to difference in suggested next time step size stored in the restart file. If #476 is merged this PR would hopefully not fail. But we should check if this effects restart before merging.

**Comment by andlaus** [comment]
Thanks for the tip. any idea why happens? I thought `flow` did its time stepping completely independently...

**Comment by totto82** [comment]
Maybe dtNext is outputted before is overwritten (set) by code in opm-simulators? If that is the case the current output of dtNext is maybe not correct either. I guess it is easy to test, just check if the value in OPM-extra is what the timestepper in ewoms suggest or the one in flow. This used to work, and should have been tested in the restart tests, so alternatively this PR messes up the ordering of the setNextTimeStepLenght(...) between code in opm-simulators and ewoms. Anyway it should not be hard to fix.

**Comment by andlaus** [comment]
thanks. I'll have look.

**Comment by andlaus** [comment]
jenkins build this opm-simulators=1731 please

**Comment by andlaus** [comment]
I slowly get the impression that the reference results are incorrect here. can anybody else have a look?

**Comment by andlaus** [comment]
> This used to work, and should have been tested in the restart tests

the restart tests only should be affected if dtNext is smaller than the length of the next report step. I guess that for the tests this was neither the case for plain-ebos nor flow.

**Comment by alfbr** [comment]
Does this impact flow in any way? We needed to store the length of the next time step in order to successfully converge after a restart. Storing it the same location as Eclipse was necessary to use the same file format. This is rather critical, so please inform if you are changing this in any way. I am out of office this week, but if you need to dig out what was done for flow, then @GitPaean probably knows.

**Comment by andlaus** [comment]
> Does this impact flow in any way?

as with #475, it shouldn't because `flow` implements its own time stepper and newton method. That said, it obviously does influence the results because some tests fail. I bet a beer that the reason for both of these failures is somewhere the output code (it is quite a beast). Maybe it even is a bug in the current code, so far nobody nows...

**Comment by totto82** [comment]
I will test this.

**Comment by totto82** [comment]
All tests pass on my machine (without opm-simulators=1731) I also didn't rebase the branch on top of current master (i.e. it uses opm restart files as default) Lets see what jenkins says.

**Comment by andlaus** [comment]
@totto82: I suffer from the same symptoms. (but not all tests pass on my machine, not even on master)

I stared a bit more on what Jenkins has to say here, and I became more and more convinced that the current reference data is wrong: the failing `OPMEXTRA`output field supposedly contains the size of the next time step, right?

**Comment by andlaus** [comment]
jenkins build this update_data please

**Comment by alfbr** [comment]
OPMEXTRA I believe is only for the old restart format, with Eclipse compatible restart it will not be written out and the size of the next time step is stored in a header. @totto82 , Torbjørn is in office and can probably help you if you need to know where it is stored and read in now.

**Comment by andlaus** [comment]
> OPMEXTRA I believe is only for the old restart format, with Eclipse compatible restart it will not be written out and the size of the next time step is stored in a header.

maybe. but if it is not supposed to be there, why does Jenkins still complain about it?

**Comment by andlaus** [comment]
jenkins build this opm-simulators=1731 update_data please

**Comment by totto82** [comment]
> why does Jenkins still complain about it


The reason is that enable-opm-rst-file=true is hardcoded in the tests. That is probably why Jenkins didn't fail when default was changed.

**Comment by andlaus** [comment]
@totto82 thanks that explains why #390 did not require a data update: it did not change anything w.r.t. the tests

**Comment by andlaus** [comment]
jenkins build this opm-simulators=1731 opm-tests=100 please

**Comment by andlaus** [comment]
*yay*

**Comment by totto82** [comment]
I would like to understand why we need to update the referance solution for this before it is merged. I will do it tomorrow.

**Comment by joakim-hove** [comment]
I would like to merge this family: 
```txt
opm-tests=97 opm-simulators=1732
```
Is that OK for you; or will it create trouble in your ongoing `update_data` endeavours?

**Comment by andlaus** [comment]
go ahead, we can tell jenkins to update the refdata again...

**Comment by andlaus** [comment]
jenkins build this opm-simulators=1731 update_data please

**Comment by andlaus** [comment]
jenkins build this opm-simulators=1731 opm-tests=100 please

File: ebos/eclproblem.hh [hunk 1]
```diff
@@ -139,6 +139,11 @@ NEW_PROP_TAG(EnableThermalFluxBoundaries);
 // The class which deals with ECL aquifers
 NEW_PROP_TAG(EclAquiferModel);
 
+NEW_PROP_TAG(MaxTimeStepSizeAfterWellEvent);
+NEW_PROP_TAG(RestartShrinkFactor);
+NEW_PROP_TAG(MaxFails);
+NEW_PROP_TAG(EnableEclTuning);
+
```
Comment (totto82) [inline ebos/eclproblem.hh]: These new parameters should be hidden from flow.

File: ebos/eclproblem.hh [hunk 2]
```diff
@@ -547,14 +577,26 @@ public:
...
+        // the "NOGRAV" keyword from Frontsim or setting the EnableGravity to false
+        // disables gravity, else the standard value of the gravity constant at sea level
+        // on earth is used
         this->gravity_ = 0.0;
-
-        // the "NOGRAV" keyword from Frontsim disables gravity...
         const auto& deck = simulator.vanguard().deck();
-        if (!deck.hasKeyword("NOGRAV") && EWOMS_GET_PARAM(TypeTag, bool, EnableGravity))
+        if (!deck.hasKeyword("NOGRAV") || !EWOMS_GET_PARAM(TypeTag, bool, EnableGravity))
             this->gravity_[dim - 1] = 9.80665;
```
Comment (totto82) [inline ebos/eclproblem.hh]: Good catch

**Comment by andlaus** [comment]
jenkins build this please

**Comment by andlaus** [comment]
jenkins build this please

File: ebos/eclproblem.hh (hunk 1 above)
Comment (andlaus) [reply ebos/eclproblem.hh]: right. I'm thinking about adding an `Ecl` prefix as well?!

Comment (totto82) [reply ebos/eclproblem.hh]: Ok, then I will wait with merging this.

**Comment by andlaus** [comment]
okay. the parameters have been renamed and OPM/opm-simulators#1733 hides them from `flow` users.

**Comment by andlaus** [comment]
jenkins build this opm-simulators=1733 please

**Comment by andlaus** [comment]
jenkins build this please

**Comment by andlaus** [comment]
I hope this was the last rebase that is necessary before merge: this PR adhered to Murphy's law in the sense that it conflicted with every single PR which was merged into master since it was opened ;)

**Pull request merged**